tests/components/esphome/test_manager.py

```diff
@@ -324,6 +325,330 @@ async def _mock_service(call: ServiceCall) -> None:
+        mock_client=mock_client,
+        device_info={"esphome_version": "2023.3.0"},
+        entry=mock_config_entry,
+    )
+    await hass.async_block_till_done()
+
+    # Register a service that returns a response
+    async def _mock_service_with_response(call: ServiceCall) -> dict[str, Any]:
+        return {"temperature": 23.5, "humidity": 65}
+
+    hass.services.async_register(
+        DOMAIN, "get_data", _mock_service_with_response, supports_response=True
+    )
+
+    # Mock the send_homeassistant_action_response method
+    mock_client.send_homeassistant_action_response = Mock()
+
+    # Call service with response template
+    device.mock_service_call(
+        HomeassistantServiceCall(
+            service="esphome.get_data",
+            data={},
+            call_id=456,
+            wants_response=True,
+            response_template="{{ response.temperature }}",
+        )
+    )
+    await hass.async_block_till_done()
+
+    # Verify response was sent back with template applied
+    mock_client.send_homeassistant_action_response.assert_called_once()
+    call_args = mock_client.send_homeassistant_action_response.call_args
+    assert call_args[0][0] == 456  # call_id
+    assert call_args[0][1] is True  # success
+    assert call_args[0][2] == ""  # error_message
+    assert call_args[0][3] == b'{"response":23.5}'
+
+
+async def test_esphome_device_service_call_with_response_template_error(
+    hass: HomeAssistant,
+    mock_config_entry: MockConfigEntry,
+    mock_client: APIClient,
+    mock_esphome_device: MockESPHomeDeviceType,
+    caplog: pytest.LogCaptureFixture,
+) -> None:
+    """Test service call with invalid response template."""
+    hass.config_entries.async_update_entry(
+        mock_config_entry, options={CONF_ALLOW_SERVICE_CALLS: True}
+    )
+    device = await mock_esphome_device(
+        mock_client=mock_client,
+        device_info={"esphome_version": "2023.3.0"},
+        entry=mock_config_entry,
+    )
+    await hass.async_block_till_done()
+
+    # Register a service that returns a response
+    async def _mock_service_with_response(call: ServiceCall) -> dict[str, Any]:
+        return {"temperature": 23.5}
+
+    hass.services.async_register(
+        DOMAIN, "get_data", _mock_service_with_response, supports_response=True
+    )
+
+    # Mock the send_homeassistant_action_response method
+    mock_client.send_homeassistant_action_response = Mock()
+
+    # Call service with invalid response template
+    device.mock_service_call(
+        HomeassistantServiceCall(
+            service="esphome.get_data",
+            data={},
+            call_id=789,
+            wants_response=True,
+            response_template="{{ response.invalid_field }}",
+        )
+    )
+    await hass.async_block_till_done()
+
+    # Verify error response was sent back
+    mock_client.send_homeassistant_action_response.assert_called_once()
+    call_args = mock_client.send_homeassistant_action_response.call_args
```

Comment:
nit: would be nice to make a helper or fixture to unpack this so you got typed args back as 0,1,2,3 is a bit hard to follow